File: 2024/11/prog_42584_주식가격.py
'''
stack = [], cur = 3 // => // stack = [3]
stack = [3], cur = 2, top = 3 // => // stack = [2]
stack = [2], cur = 3, top = 2 // => // stack = [2] *** 
stack = [2], cur = 2, top = 2 // => // stack = [2]
stack = [2], cur = 1, top = 2 // => // stack = [1] *** 
'''

def solution(prices):
    answer = []
    stack = []
    prices = prices[::-1]
    
    for i in range(len(prices)):
        cur = prices[i]
        is_find = False
        
        while stack:
            top, idx = stack.pop()

            if cur > top:
                answer.append(i - idx)
                stack.append((top, idx))
                is_find = True
                break
                
        stack.append((cur, i))
                
        if not is_find: # 끝까지 가격 안 떨어짐 
            answer.append(i)

    return answer[::-1]

File: 2025/07/prog_42584_주식가격.js
function solution(prices) {
    var answer = []
    for (let i = prices.length-1; i > -1; i--) {
        answer.push(i)
    }
    
    let stack = []
    
    for (let curI = prices.length-1; curI > -1; curI--) {
        const cur = prices[curI]
        let isFind = false
        
        // 스택의 각 값 탐색 
        while (stack.length > 0) {
            let [top, topI] = stack[stack.length-1] 
            
            if (cur <= top) {
                stack.pop()
            } else if (cur > top) {
                answer[curI] = (topI - curI)
                isFind = true
                break
            } 
        }

        stack.push([cur, curI])
    }

    return answer;
}